src/SessionSetup/RetrieveInstancesWidget.cpp

```diff
@@ -4,27 +4,160 @@
+  s_initial_loading_failed_.assignProperty(ui_->allCheckBox, "enabled", false);
+  s_initial_loading_failed_.addTransition(this, &RetrieveInstancesWidget::LoadingStarted,
+                                          &s_loading_);
+}
+
+std::optional<Project> RetrieveInstancesWidget::GetQSettingsProject() {
+  QSettings settings;
+
+  std::optional<Project> remembered_project;
+  if (settings.contains(kSelectedProjectIdKey)) {
+    remembered_project = Project{
+        settings.value(kSelectedProjectDisplayNameKey).toString() /* display_name */,
+        settings.value(kSelectedProjectIdKey).toString() /* id */
+    };
+  }
+  return remembered_project;
+}
+
+void RetrieveInstancesWidget::Start() {
+  state_machine_.setInitialState(&s_loading_);
+  state_machine_.start();
+
+  QSettings settings;
+  ui_->allCheckBox->setChecked(settings.value(kAllInstancesKey, false).toBool());
+
+  InitialLoad(GetQSettingsProject());
+}
+
+Client::InstanceListScope RetrieveInstancesWidget::GetInstanceListScope() {
+  return ui_->allCheckBox->isChecked() ? Client::InstanceListScope::kAllReservedInstances
+                                       : Client::InstanceListScope::kOnlyOwnInstances;
+}
+
+void RetrieveInstancesWidget::InitialLoad(const std::optional<Project>& remembered_project) {
+  emit LoadingStarted();
+  retreive_instances_->LoadProjectsAndInstances(remembered_project, GetInstanceListScope())
+      .Then(main_thread_executor_,
+            [widget = QPointer<RetrieveInstancesWidget>(this)](
+                ErrorMessageOr<RetrieveInstances::LoadProjectsAndInstancesResult> loading_result) {
+              if (widget == nullptr) return;
+
+              if (loading_result.has_error()) {
+                emit widget->InitialLoadingFailed();
+                widget->OnInstancesLoadingReturned(loading_result.error());
+                return;
+              }
+
+              CHECK(widget->ui_->projectComboBox->count() == 0);
+
+              RetrieveInstances::LoadProjectsAndInstancesResult& data{loading_result.value()};
+
+              QString default_project_text =
+                  QString("Default Project (%1)").arg(data.default_project.display_name);
+              widget->ui_->projectComboBox->addItem(default_project_text);
+
+              QVector<Project>& projects{data.projects};
+              std::sort(projects.begin(), projects.end(), [](const Project& p1, const Project& p2) {
+                return p1.display_name.toLower() < p2.display_name.toLower();
+              });
+
+              for (const Project& project : data.projects) {
+                QString text = project.display_name;
+                if (project == data.default_project) {
+                  text.append(" (default)");
+                }
+                widget->ui_->projectComboBox->addItem(text, QVariant::fromValue(project));
+
+                if (project == data.project_of_instances) {
```

Comment:
I don't understand why this is needed, could you add a comment?

Reply:
I added a comment here and I added some comments to the RetrieveInstances. ptal